src/mlpack/core/metrics/bleu_impl.hpp

```diff
@@ -0,0 +1,200 @@
+  std::vector<size_t> possibleMatchesByOrder(maxOrder, 0);
+
+  // referenceLength: It is the sum of minimum length of the paragraph from
+  // various documents.
+  // translationLength: It is the sum of length of each paragraphs.
+  referenceLength = 0, translationLength = 0;
+
+  auto refIt = referenceCorpus.cbegin();
+  auto trIt = translationCorpus.cbegin();
+  for (; refIt != referenceCorpus.cend(), trIt != translationCorpus.cend();
+      ++refIt, ++trIt)
+  {
+    size_t min = std::numeric_limits<size_t>::max();
+    for (const auto& t : *refIt)
+    {
+      if (min > t.size())
+      {
+        min = t.size();
+      }
+    }
+
+    if (min == std::numeric_limits<size_t>::max())
+      min = 0;
+
+    referenceLength += min;
+    translationLength += trIt->size();
+
+    // mergedRefNGramCounts: It accumulates all the similar n-grams from
+    // various references or documents, so that there is no repetition of
+    // any key (sequence of order n).
+    std::map<WordVector, size_t> mergedRefNGramCounts;
+    for (const auto& t : *refIt)
+    {
+      // ngram: It holds the n-grams of each document/reference.
+      const std::map<WordVector, size_t> ngrams = GetNGrams(t);
+      for (auto it = ngrams.cbegin(); it != ngrams.cend(); ++it)
+      {
+        mergedRefNGramCounts[it->first] = std::max(it->second,
+            mergedRefNGramCounts[it->first]);
+      }
+    }
+    // translationNGramCounts: It extracts the n-grams of the generated text
+    // sequence.
+    const std::map<WordVector, size_t> translationNGramCounts
+        = GetNGrams(*trIt);
+
+    // overlap: It holds those key (sequence of order n) which are common to
+    // reference corpus and translation corpus.
+    std::map<WordVector, size_t> overlap;
+    for (auto it = translationNGramCounts.cbegin();
+         it != translationNGramCounts.cend();
+         ++it)
+    {
+      if (mergedRefNGramCounts.find(it->first) != mergedRefNGramCounts.end())
+      {
+        // If the key (sequence of order n) is present in both translation
+        // corpus as well as reference corpus, then the minimum number of
+        // counts it has occurred in any is considered.
+        overlap[it->first] = std::min(mergedRefNGramCounts.count(it->first),
+                                      it->second);
+      }
+    }
+
+    for (auto it = overlap.cbegin(); it != overlap.cend(); ++it)
+    {
+      matchesByOrder[it->first.size() - 1] += it->second;
+    }
+
+    for (size_t order = 1; order < maxOrder + 1; ++order)
+    {
+      if (order < trIt->size() + 1)
+        possibleMatchesByOrder[order - 1] += trIt->size() - order + 1;
+    }
+  }
+
+  precisions = PrecisionType(maxOrder, 0.0);
+
+  if (smooth)
+  {
+    for (size_t i = 0; i < maxOrder; ++i)
+    {
+      precisions[i]
+          = (matchesByOrder[i] + 1.0) / (possibleMatchesByOrder[i] + 1.0);
+    }
+  }
+  else
+  {
+    for (size_t i = 0; i < maxOrder; ++i)
+    {
+      if (possibleMatchesByOrder[i] > 0.0)
```

Comment:
There's no need to convert to `double`, `possibleMatchesByOrder[i]` is an integer.
```suggestion
      if (possibleMatchesByOrder[i] > 0)
```